colony_picker/inverse_kinematics.py

```diff
@@ -0,0 +1,634 @@
+            [alpha 1, alpha 2, ... , alpha n],
+            [a 1, a 2, ... , a n], 
+            [d 1, d 2, ... , d n]]
+            where n is the number of joints on the robot arm.
+        smart_seed (bool, optional): whether the algorithm should randomly seed
+            the optimization until an error lower than the tolerance is achieved
+            if the error is too high. Defaults to False.
+        tolerance (float, optional): The error tolerated between the desired
+            end effector pose and the end effector pose calculated from the
+            joint values the optimization solved for. Defaults to 1e-06.
+        seeding_max_iters (int, optional): the number of times to optimize again
+            with a new random set of thetas_init before giving up on achieving
+            the error tolerance. Defaults to 100.
+
+    Returns:
+        Tuple(np.ndarray, bool): the set of joint angles the inverse kinematics
+            algorithm solved for and whether the joint angles achieve the 
+            desired end effector pose within the error tolerance in the format
+            (joint angles, whether algorithm achieved error within tolerance)
+    """
+    solved = False
+    result = minimize(objective_function, thetas_init,
+                      (desired_end_effector_pose, dh_params,), method='BFGS')
+    error = result.fun
+    if error <= tolerance:
+        solved = True
+    joint_angles = result.x
+
+    # trying new random seeds until error is within tolerance
+    if not solved:
+        if smart_seed:
+            num_joints = len(dh_params[0])
+            for i in range(seeding_max_iters):
+                new_seed = np.array([random.uniform(-math.pi, math.pi)
+                                    for j in range(num_joints)])
+                new_result = minimize(objective_function, new_seed,
+                                      (desired_end_effector_pose, dh_params,),
+                                      method='BFGS')
+                new_joint_angles = new_result.x
+                new_error = new_result.fun
+                if new_error < error:
+                    error = new_error
+                    joint_angles = new_joint_angles
+                if new_error <= tolerance:
+                    solved = True
+                    break
+    return (joint_angles, solved)
+
+# ************************************** #
+#     HELPER FUNCTIONS FOR ANIMATIONS    #
+# ************************************** #
+
+
+def draw_coordinate_system(plotter: pv.Plotter, t_mat: np.ndarray,
+                           base: bool = False, name: str = ""):
+    """draws a coordinate system on the plotter with the orientation and
+    position described by t_mat and with the x, y, and z axes denoted by red,
+    green, and blue colors. 
+
+    Args:
+        plotter (pv.Plotter): the plotter object to draw the coordinate system
+            on.
+        t_mat (np.ndarray): a transformation matrix describing the position and
```

Comment:
Not sure if it's useful to say `4x4 transformation matrix` or if it's implied by the name